src/test/clojurescript/clara/test_salience.cljs

```diff
@@ -0,0 +1,115 @@
+(ns clara.test-salience
+  (:require-macros [cljs.test :refer (is deftest run-tests testing)]
+    [clara.test-rules-data])
+  (:require [cljs.test :as t]
+    [clara.rules.engine :as eng]
+    [clara.rules.accumulators :as acc]
+    [clara.rules :refer [insert retract fire-rules query insert!]
+     :refer-macros [defrule defsession defquery]]
+    [clara.rules.testfacts :refer [->Temperature Temperature
+                                   ->WindSpeed WindSpeed
+                                   ->ColdAndWindy ColdAndWindy]]))
+
+(def salience-rule-output (atom []))
+
+(defrule salience-rule1
+  {:salience 100}
+  [Temperature]
+  =>
+  (swap! salience-rule-output conj 100))
+
+(defrule salience-rule2
+  {:salience 50}
+  [Temperature]
+  =>
+  (swap! salience-rule-output conj 50))
+
+(defrule salience-rule3
+  {:salience 0}
+  [Temperature]
+  =>
+  (swap! salience-rule-output conj 0))
+
+(defrule salience-rule4
+  {:salience -50}
+  [Temperature]
+  =>
+  (swap! salience-rule-output conj -50))
+
+
+(deftest test-salience
+  (doseq [[sort-fn
+           group-fn
+           expected-order]
+
+          [[:default-sort :default-group :forward-order]
+           [:default-sort :salience-group :forward-order]
+           [:default-sort :neg-salience-group :backward-order]
+
+           [:numeric-greatest-sort :default-group :forward-order]
+           [:numeric-greatest-sort :salience-group :forward-order]
+           [:numeric-greatest-sort :neg-salience-group :backward-order]
+
+
+           [:boolean-greatest-sort :default-group :forward-order]
+           [:boolean-greatest-sort :salience-group :forward-order]
+           [:boolean-greatest-sort :neg-salience-group :backward-order]
+
+
+           [:numeric-least-sort :default-group :backward-order]
+           [:numeric-least-sort :salience-group :backward-order]
+           [:numeric-least-sort :neg-salience-group :forward-order]
+
+           [:boolean-least-sort :default-group :backward-order]
+           [:boolean-least-sort :salience-group :backward-order]
+           [:boolean-least-sort :neg-salience-group :forward-order]]]
+
+    (let [numeric-greatest-sort (fn [x y]
+                                  (cond
+                                    (= x y) 0
+                                    (> x y) -1
+                                    :else 1))
+
+          numeric-least-sort (fn [x y]
+                               (numeric-greatest-sort y x))
+
+          salience-group-fn (fn [production]
+                              (or (some-> production :props :salience)
+                                0))
+
+          neg-salience-group-fn (fn [p]
+                                  (- (salience-group-fn p)))]
+
+      (defsession test-salience-session 'clara.test-salience
```

Comment:
Defsession exists since rulebase compilation in ClojureScript has to happen at compile time, rather than at runtime.  Creating sessions at runtime in Clojure makes heavy use of eval, which was historically absent in ClojureScript.  I know there has been some work on adding eval to ClojureScript; it would be nice if that was mature enough to use but I haven't looked into it enough to assess that.

I think what's happening here is that the actual rule network construction is being done at compilation time, and the macro is working out so that the functionality controlled by the options, which isn't reliant on eval, gets done at runtime.  This seems to work, but is a little odd.  I'm not sure what we should do here, I'll have to give it some thought.  Thoughts @rbrush ?